momepy/functional/_intensity.py

```diff
@@ -48,3 +50,67 @@ def _calculate_courtyards(group):
     )
 
     return result
+
+
+def node_density(
+    left: GeoDataFrame, right: GeoDataFrame, graph: Graph, weighted: bool = False
+) -> Series:
+    """Calculate the density of a node's neighbours (for all nodes)
+    on the street network defined in ``graph``.
+
+    Calculated as the number of neighbouring
+    nodes / cumulative length of street network within neighbours.
+    ``node_start``,  ``node_end``, is standard output of
+    :py:func:`momepy.nx_to_gdf` and is compulsory for ``right`` to have
+    these columns.
+    If ``weighted``, a ``degree`` column is also required in ``left``.
+
+    Adapted from :cite:`dibble2017`.
+
+    Parameters
+    ----------
+    left : GeoDataFrame
+        A GeoDataFrame containing nodes of street network.
+    right : GeoDataFrame
+        A GeoDataFrame containing edges of street network.
+    graph :  libpysal.graph.Graph
+        A spatial weights matrix capturing relationship between nodes.
+    weighted : bool (default False)
+        If ``True``, density will take into account node degree as ``k-1``.
+
+    Returns
+    -------
+    Series
+        A Series containing resulting values.
+
+    Examples
+    --------
+    >>> nodes['density'] = mm.node_density(nodes, edges, graph)
+    """
+
+    required_cols = ["node_start", "node_end"]
+    if not np.isin(required_cols, right.columns).all():
+        raise ValueError(
+            f"Columns { *required_cols, } are needed in "
+            "right GeoDataframe for the calculations."
+        )
+    if weighted and ("degree" not in left.columns):
+        raise ValueError(
+            "Degree column is needed in " "left GeoDataframe for density calculations."
```

Comment:
This has weird formatting. We should also be explicit by saying `"degree"` rather than `Degree`.